Update inputs from DocDB

sbailey opened the pull request:
This PR simplifies and documents the process of updating inputs from DocDB.  Updating all the inputs started getting out of hand so I'd like to review and merge this PR to get the basic structure established before moving on with other inputs so that this doesn't become a super mega PR.

This PR goes along with the update_inputs branch in desimodel svn.

Included here:
  * updated throughputs from [DESI-0347v11](https://desi.lbl.gov/DocDB/cgi-bin/private/ShowDocument?docid=347&version=11) (instead of v5)
  * updated the positioner locations from [DESI-0530v5](https://desi.lbl.gov/DocDB/cgi-bin/private/ShowDocument?docid=530&version=10) (instead of the now deprecated DESI-0403v2)
  * updated positioner:fiber mapping following the cassette:bundle mapping in
    [DESI-2721v2](https://desi.lbl.gov/DocDB/cgi-bin/private/ShowDocument?docid=2721&version=2)

Not included here:
  * updating blur and offset parameters; these come from `data/inputs/throughput/raytracing.txt` which we got from the engineering team and we keep asking to have added to DocDB, but it still doesn't seem to be there
  * updating PSF model
  * refactoring how fiber input losses are parameterized and used

Structurally this PR also refactors how the inputs from DocDB are downloaded and propagated.  e.g. updating the throughput parameters is now done via:
```python
import desimodel.inputs.throughput
desimodel.inputs.throughput.update()
```
That will download the following files directly from DocDB:
  * From DESI-0347v11: `DESI-347-v11 Throughput Noise SNR Calcs.xlsx` and `desi.yaml`
  * From DESI-0334v3: `blue-thru.txt`, `red-thru.txt`, and `nir-thru-250.txt`
It then uses the `xlrd` module to directly parse the Excel spreadsheet to extract the necessary columns rather than relying upon download-then-cut-and-paste-by-hand to get the necessary rows/columns into a particular text format before further processing.

As a result, the following are deprecated and removed:
  * `bin/combine_throughputs.py` (now in `desimodel.inputs.throughput.update()`
  * `bin/randomize_fibers.py` (now in `desimodel.inputs.fiberpos.update()`
  * Most files in svn `data/inputs/`: those were just intermediate files from DocDB (sometimes directly
  from DocDB, sometimes requiring manual cut-and-pasting from Excel).  Now all parsing is done directly
  on the originals downloaded from DocDB. These are cached in `data/inputs/docdb/` but *not*
  added to svn.

Other notes:
  * refactoring how to download and directly parse the Excel files was largely motivated by an update not yet included here: the spectrograph throughput numbers from DESI-0347 now have high resolution measurements of the as-built parts from vendor testing, but these are split across non-contiguous columns on different sheets spread across ~7 different Excel files and the cutting-and-pasting by hand was going to be fragile.
  * `desimodel.inputs.*` requires python3 and the `xlrd` module to parse Excel files.  This is only needed to update the inputs from DocDB into svn and are not otherwise needed to use desimodel.
  * Thanks to @sybenzvi for pointing me to the pandas code and thus `xlrd` which makes parsing these Excel files pretty easy

I'd like to do more sanity checks myself before merging this, but I'm starting this PR now to get other eyes on it and review the structure and basic philosophy.  In particular, can @dkirkby take a look at this, since you are the other person who has recently dug into the details of DocDB vs. desimodel parameter files?

dkirkby commented on py/desimodel/inputs/throughput.py:
```diff
@@ -0,0 +1,149 @@
+    wave = docdb.xls_read_row(filename, 'Throughput', 3, 'C', 'P')*10
+    thru = docdb.xls_read_row(filename, 'Throughput', 112, 'C', 'P')
+    specthru = docdb.xls_read_row(filename, 'Throughput', 93, 'C', 'P')
+
```
Since the excel file is hand maintained, we could get burned by hard coding the cell ranges like this in case they change or typos are introduced.  Adding some sanity checks on the extracted arrays would offer some protection, e.g., `assert np.all(np.diff(wave) > 0)`.

dkirkby commented:
As part of this cleanup, I would like to update the fiberloss files in `data/throughput/` to the results from the latest GalSim calculations, and remove / deprecate the original `bin/fiberloss.py`.   The main difference will be from considering a range of galaxy shapes, rather than the round edge case, which will generally lower the fiberloss values.  Specifically, I would use the medians of the families of curves plotted for QSO, ELG, LRG in DESI-2720 with both instrumental & astrophysical effects enabled (e.g., p.15).

moustakas commented:
Could we please also add a fiducial fiberloss file for BGS targets?

dkirkby commented:
@moustakas Yes!  I just need a fiducial model (e.g., mixture of Sersic indices) and priors on the size and shape parameters.

sbailey commented:
I will add sanity checks on the numbers extracted from the excel files, including reading their column headers and verifying that they haven't changed.  Also note that the code grabs a fixed version of the file, not the latest, so updating versions requires a proactive change and reviewing what rows/columns to include should be part of vetting that update.  But I agree that basic sanity checks will help protect against errors.

> As part of this cleanup, I would like to update the fiberloss files in data/throughput/ to the results from the latest GalSim calculations, and remove / deprecate the original bin/fiberloss.py. The main difference will be from considering a range of galaxy shapes, rather than the round edge case, which will generally lower the fiberloss values. Specifically, I would use the medians of the families of curves plotted for QSO, ELG, LRG in DESI-2720 with both instrumental & astrophysical effects enabled (e.g., p.15).

For this PR, let's
* update `data/throughput/fiberloss-*.dat` with new numbers in the same format
* add `data/throughput/fiberloss-bgs.dat`
* remove `bin/fiberloss.py` and replace it with code in `py/desimodel/fiberloss.py` (wrapping specsim) and/or rst docs or a jupyter notebook showing how to call that to generate the fiberloss files
* Optional for this PR: also add the multi-object fiberloss realizations vs. x,y,seeing,etc

i.e. we're moving toward deprecating "one fiberloss per object type" and replacing it with galsim-on-the-fly or pre-calculated realizations, but let's not break backwards compatibility yet.

dkirkby commented:
> update data/throughput/fiberloss-*.dat with new numbers in the same format

Ok if these are in ECSV, with units, etc, or do you prefer the current plain txt format?

> add data/throughput/fiberloss-bgs.dat

Fine with me, but I will need input from @moustakas so he may want to push this to a later PR.

> remove bin/fiberloss.py and replace it with...

I propose a streamlined version of this notebook https://github.com/desihub/specsim/blob/master/docs/nb/FiberlossExamples.ipynb

> Optional for this PR: also add the multi-object fiberloss realizations vs. x,y,seeing,etc

Lets do that separately since it needs some complex sampling logic that we haven't fleshed out yet.

> ...but let's not break backwards compatibility yet.

Agreed. I have gone to some trouble to keep `specsim` backwards compatible :-)

sbailey commented:
Strictly preserving backwards compatibility for this PR implies preserving the format of `fiberloss-*.dat`, i.e. not going to ECSV with units.  However, I only use these files as an intermediate step on the way to the `thru-*.fits` files, so updating their format is ok for me; any impact would be handled internal to desimodel.  IIUC you can handle this on the specsim side by just changing the fiberloss.table.format to 'ascii.ecsv' in `desi.config`.

The approach of "one fiberloss per object type" is useful in the spirit of quicksim but fundamentally wrong in detail so we shouldn't go to much effort to upgrade its format if the intention is to throw it away soon anyway.  i.e. I'm also fine with the existing format, while focussing on ECSV conversions for things that will stick around.

sbailey commented:
I added the sanity checks on which rows/columns were extracted from the DocDB excel files.  This was useful, catching a row shift in DESI-0530 from v10 to v11.

I also updated the data files in svn branch update_inputs:

https://desi.lbl.gov/trac/browser/code/desimodel/branches/update_inputs/data/focalplane?order=name
https://desi.lbl.gov/trac/browser/code/desimodel/branches/update_inputs/data/throughput

Full disclosure — this includes some non-backwards compatible changes that I could back out if they are problematic:

`data/focalplane/fiberpos.txt` column "positioner" renamed "FPDEVICE" (focal plane device) for consistency with the hardware team naming scheme in DESI-2724.  For them, "positioner" refers to the serial number of the positioner object itself, not the hole into which it was placed.  The hole is referred to as a "device" with range [0-542] and resets to 0 with each new petal.  `fpdevice = 1000*petal + device` fulfills the role of what we previously called "positioner", though we are encouraged to switch to reporting focal plane locations as (petal, device).

--> fiberassign is the only code that I know of that uses that file, and it parses it without reading column names so we should be fine.

`data/focalplane/fiberpos.fits` gained additional columns, switched to uppercase column names, renamed "positioner" -> "FPDEVICE", and renamed "spectrograph" -> "SPECTRO" for symmetry with DESI-2724 and the text file.

--> This will require a minor fix to `desisim.targets.get_targets`; irritating, but it's imminent for a refactor anyway.  I could add a hook to `desimodel.io.load_fiberpos()` to temporarily support both `SPECTRO` and `spectrograph` to ease the transition.

`data/focalplane/fiberpos-all.[fits, ecsv]` are new files that include all focal plane locations including sky monitor fibers and fidicials, not just the positioners that have fibers going to the spectrographs.

Going to check on those Travis failures now...

sbailey commented:
Ugh.  Travis tests are failing due to testing against lightweight version of desimodel data, that have the old column names.  Stepping away from the keyboard to reconsider...

sbailey commented:
To maintain backwards compatibility for this PR, I've updated `desimodel.io.load_fiberpos` to print a deprecation warning if it finds an old `fiberpos.fits` file, while still supporting both FPDEVICE/POSITIONER and SPECTRO/SPECTOGRAPH names.  We can remove this after desisim and desimodel lite test have been updated.

In a separate chat with @dkirkby we agreed to make the fiberloss updates as part of a separate PR to not block the fiberpos and throughput updates here.

I'm ready to merge.  Last call for comments on these changes.
